// eslint-plugin-sort-class-members/dist/rules/sort-class-members.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.sortClassMembers = undefined;

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _schema = require('./schema');

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var sortClassMembers = exports.sortClassMembers = {
	getRule: function getRule() {
		var defaults = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

		function sortClassMembersRule(context) {
			var options = context.options[0] || {};
			var stopAfterFirst = !!options.stopAfterFirstProblem;
			var accessorPairPositioning = options.accessorPairPositioning || 'getThenSet';
			var order = options.order || defaults.order || [];
			var groups = _extends({}, builtInGroups, defaults.groups, options.groups);
			var orderedSlots = getExpectedOrder(order, groups);
			var groupAccessors = accessorPairPositioning !== 'any';

			return {
				ClassDeclaration: function ClassDeclaration(node) {
					var members = getClassMemberInfos(node, context.getSourceCode(), orderedSlots);

					// check for out-of-order and separated get/set pairs
					var accessorPairProblems = findAccessorPairProblems(members, accessorPairPositioning);
					var _iteratorNormalCompletion = true;
					var _didIteratorError = false;
					var _iteratorError = undefined;

					try {
						for (var _iterator = accessorPairProblems[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
							var problem = _step.value;

							var message = 'Expected {{ source }} to come immediately {{ expected }} {{ target }}.';

							reportProblem({ problem: problem, context: context, message: message, stopAfterFirst: stopAfterFirst, problemCount: problemCount });
							if (stopAfterFirst) {
								break;
							}
						}

						// filter out the second accessor in each pair so we only detect one problem
						// for out-of-order	accessor pairs
					} catch (err) {
						_didIteratorError = true;
						_iteratorError = err;
					} finally {
						try {
							if (!_iteratorNormalCompletion && _iterator.return) {
								_iterator.return();
							}
						} finally {
							if (_didIteratorError) {
								throw _iteratorError;
							}
						}
					}

					members = members.filter(function (m) {
						return !(m.matchingAccessor && !m.isFirstAccessor);
					});

					// ignore members that don't match any slots
					members = members.filter(function (member) {
						return member.acceptableSlots.length;
					});

					// check member positions against rule order
					var problems = findProblems(members);
					var problemCount = problems.length;
					var _iteratorNormalCompletion2 = true;
					var _didIteratorError2 = false;
					var _iteratorError2 = undefined;

					try {
						for (var _iterator2 = problems[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
							var _problem = _step2.value;

							var _message = 'Expected {{ source }} to come {{ expected }} {{ target }}.';
							reportProblem({
								problem: _problem,
								message: _message,
								context: context,
								stopAfterFirst: stopAfterFirst,
								problemCount: problemCount,
								groupAccessors: groupAccessors
							});

							if (stopAfterFirst) {
								break;
							}
						}
					} catch (err) {
						_didIteratorError2 = true;
						_iteratorError2 = err;
					} finally {
						try {
							if (!_iteratorNormalCompletion2 && _iterator2.return) {
								_iterator2.return();
							}
						} finally {
							if (_didIteratorError2) {
								throw _iteratorError2;
							}
						}
					}
				}
			};
		}

		sortClassMembersRule.schema = _schema.sortClassMembersSchema;

		return sortClassMembersRule;
	}
};

function reportProblem(_ref) {
	var problem = _ref.problem,
	    message = _ref.message,
	    context = _ref.context,
	    stopAfterFirst = _ref.stopAfterFirst,
	    problemCount = _ref.problemCount,
	    groupAccessors = _ref.groupAccessors;
	var source = problem.source,
	    target = problem.target,
	    expected = problem.expected;

	var reportData = {
		source: getMemberDescription(source, { groupAccessors: groupAccessors }),
		target: getMemberDescription(target, { groupAccessors: groupAccessors }),
		expected: expected
	};

	if (stopAfterFirst && problemCount > 1) {
		message += ' ({{ more }} similar {{ problem }} in this class)';
		reportData.more = problemCount - 1;
		reportData.problem = problemCount === 2 ? 'problem' : 'problems';
	}

	context.report({ node: source.node, message: message, data: reportData });
}

function getMemberDescription(member, _ref2) {
	var groupAccessors = _ref2.groupAccessors;

	if (member.kind === 'constructor') {
		return 'constructor';
	}

	var typeName = void 0;
	if (member.matchingAccessor && groupAccessors) {
		typeName = 'accessor pair';
	} else if (isAccessor(member)) {
		typeName = member.kind + 'ter';
	} else {
		typeName = member.type;
	}

	return '' + (member.static ? 'static ' : '') + typeName + ' ' + member.name;
}

function getClassMemberInfos(classDeclaration, sourceCode, orderedSlots) {
	var classMemberNodes = classDeclaration.body.body;

	var members = classMemberNodes.map(function (member, i) {
		return _extends({}, getMemberInfo(member, sourceCode), { id: String(i) });
	}).map(function (memberInfo, i, memberInfos) {
		matchAccessorPairs(memberInfos);
		var acceptableSlots = getAcceptableSlots(memberInfo, orderedSlots);
		return _extends({}, memberInfo, { acceptableSlots: acceptableSlots });
	});

	return members;
}

function getMemberInfo(node, sourceCode) {
	var name = void 0;
	var type = void 0;
	var propertyType = void 0;

	if (node.type === 'ClassProperty') {
		type = 'property';

		var _sourceCode$getFirstT = sourceCode.getFirstTokens(node, 2),
		    _sourceCode$getFirstT2 = _slicedToArray(_sourceCode$getFirstT, 2),
		    first = _sourceCode$getFirstT2[0],
		    second = _sourceCode$getFirstT2[1];

		name = second && second.type === 'Identifier' ? second.value : first.value;
		propertyType = node.value ? node.value.type : node.value;
	} else {
		name = node.key.name;
		type = 'method';
	}

	return { name: name, type: type, static: node.static, kind: node.kind, propertyType: propertyType, node: node };
}

function findAccessorPairProblems(members, positioning) {
	var problems = [];
	if (positioning === 'any') {
		return problems;
	}

	forEachPair(members, function (first, second, firstIndex, secondIndex) {
		if (first.matchingAccessor === second.id) {
			var outOfOrder = positioning === 'getThenSet' && first.kind !== 'get' || positioning === 'setThenGet' && first.kind !== 'set';
			var outOfPosition = secondIndex - firstIndex !== 1;

			if (outOfOrder || outOfPosition) {
				var expected = outOfOrder ? 'before' : 'after';
				problems.push({ source: second, target: first, expected: expected });
			}
		}
	});

	return problems;
}

function findProblems(members) {
	var problems = [];

	forEachPair(members, function (first, second) {
		if (!areMembersInCorrectOrder(first, second)) {
			problems.push({ source: second, target: first, expected: 'before' });
		}
	});

	return problems;
}

function forEachPair(list, callback) {
	list.forEach(function (first, firstIndex) {
		list.slice(firstIndex + 1).forEach(function (second, secondIndex) {
			callback(first, second, firstIndex, firstIndex + secondIndex + 1);
		});
	});
}

function areMembersInCorrectOrder(first, second) {
	return first.acceptableSlots.some(function (a) {
		return second.acceptableSlots.some(function (b) {
			return a.index === b.index && areSlotsAlphabeticallySorted(a, b) ? first.name.localeCompare(second.name) <= 0 : a.index <= b.index;
		});
	});
}

function areSlotsAlphabeticallySorted(a, b) {
	return a.sort === 'alphabetical' && b.sort === 'alphabetical';
}

function getAcceptableSlots(memberInfo, orderedSlots) {
	return orderedSlots.map(function (slot, index) {
		return { index: index, score: scoreMember(memberInfo, slot), sort: slot.sort };
	}) // check member against each slot
	.filter(function (_ref3) {
		var score = _ref3.score;
		return score > 0;
	}) // discard slots that don't match
	.sort(function (a, b) {
		return b.score - a.score;
	}) // sort best matching slots first
	.filter(function (_ref4, i, array) {
		var score = _ref4.score;
		return score === array[0].score;
	}) // take top scoring slots
	.sort('index');
}

function scoreMember(memberInfo, slot) {
	if (!Object.keys(slot).length) {
		return 1; // default/everything-else slot
	}

	var scores = comparers.map(function (_ref5) {
		var property = _ref5.property,
		    value = _ref5.value,
		    test = _ref5.test;

		if (slot[property] !== undefined) {
			return test(memberInfo, slot) ? value : -1;
		}

		return 0;
	});

	if (scores.indexOf(-1) !== -1) {
		return -1;
	}

	return scores.reduce(function (a, b) {
		return a + b;
	});
}

function getExpectedOrder(order, groups) {
	return flatten(order.map(function (s) {
		return expandSlot(s, groups);
	}));
}

function expandSlot(input, groups) {
	if (Array.isArray(input)) {
		return input.map(function (x) {
			return expandSlot(x, groups);
		});
	}

	var slot = void 0;
	if (typeof input === 'string') {
		slot = input[0] === '[' // check for [groupName] shorthand
		? { group: input.substr(1, input.length - 2) } : { name: input };
	} else {
		slot = _extends({}, input);
	}

	if (slot.group) {
		if (groups.hasOwnProperty(slot.group)) {
			return expandSlot(groups[slot.group], groups);
		}

		// ignore undefined groups
		return [];
	}

	var testName = slot.name && getNameComparer(slot.name);
	if (testName) {
		slot.testName = testName;
	}

	return [slot];
}

function isAccessor(_ref6) {
	var kind = _ref6.kind;

	return kind === 'get' || kind === 'set';
}

function matchAccessorPairs(members) {
	forEachPair(members, function (first, second) {
		var isMatch = first.name === second.name && first.static === second.static;
		if (isAccessor(first) && isAccessor(second) && isMatch) {
			first.isFirstAccessor = true;
			first.matchingAccessor = second.id;
			second.matchingAccessor = first.id;
		}
	});
}

function getNameComparer(name) {
	if (name[0] === '/') {
		var namePattern = name.substr(1, name.length - 2);

		if (namePattern[0] !== '^') {
			namePattern = '^' + namePattern;
		}

		if (namePattern[namePattern.length - 1] !== '$') {
			namePattern += '$';
		}

		var re = new RegExp(namePattern);

		return function (n) {
			return re.test(n);
		};
	}

	return function (n) {
		return n === name;
	};
}

function flatten(collection) {
	var result = [];

	var _iteratorNormalCompletion3 = true;
	var _didIteratorError3 = false;
	var _iteratorError3 = undefined;

	try {
		for (var _iterator3 = collection[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
			var item = _step3.value;

			if (Array.isArray(item)) {
				result.push.apply(result, _toConsumableArray(flatten(item)));
			} else {
				result.push(item);
			}
		}
	} catch (err) {
		_didIteratorError3 = true;
		_iteratorError3 = err;
	} finally {
		try {
			if (!_iteratorNormalCompletion3 && _iterator3.return) {
				_iterator3.return();
			}
		} finally {
			if (_didIteratorError3) {
				throw _iteratorError3;
			}
		}
	}

	return result;
}

var builtInGroups = {
	constructor: { name: 'constructor', type: 'method' },
	properties: { type: 'property' },
	getters: { kind: 'get' },
	setters: { kind: 'set' },
	'accessor-pairs': { accessorPair: true },
	'static-properties': { type: 'property', static: true },
	'conventional-private-properties': { type: 'property', name: '/_.+/' },
	'arrow-function-properties': { propertyType: 'ArrowFunctionExpression' },
	methods: { type: 'method' },
	'static-methods': { type: 'method', static: true },
	'conventional-private-methods': { type: 'method', name: '/_.+/' },
	'everything-else': {}
};

var comparers = [{ property: 'name', value: 100, test: function test(m, s) {
		return s.testName(m.name);
	} }, { property: 'type', value: 10, test: function test(m, s) {
		return s.type === m.type;
	} }, { property: 'static', value: 10, test: function test(m, s) {
		return s.static === m.static;
	} }, { property: 'kind', value: 10, test: function test(m, s) {
		return s.kind === m.kind;
	} }, {
	property: 'accessorPair',
	value: 20,
	test: function test(m, s) {
		return s.accessorPair && m.matchingAccessor || s.accessorPair === false && !m.matchingAccessor;
	}
}, {
	property: 'propertyType',
	value: 11,
	test: function test(m, s) {
		return m.type === 'property' && s.propertyType === m.propertyType;
	}
}];
